### PR TITLE
Add idp userid to users table.

### PR DESCRIPTION
**Pull request by pwei1018**
*Issue #:* /bcgov/entity###
https://github.com/bcgov/entity/issues/14367

*Description of changes:*
1. add idp_userid column to users table;

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the lear license (Apache 2.0).

### CODE REVIEW

**Review by bolyachevets — APPROVED** [review]

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+# revision identifiers, used by Alembic.
+revision = '1ba027b074a5'
+down_revision = 'e10f724fad6d'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.add_column('users', sa.Column('idp_userid', sa.String()))
```
Comment (argush3) [inline legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py:21]: can we update the column def to the following?  length 256 is what auth user table has. 

`op.add_column('users', sa.Column('idp_userid', sa.String(length=256), nullable=True)`

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    #op.drop_constraint('users_sub_key', 'users', type_='unique')
```
Comment (seeker25) [inline legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py:22]: Any reason for the commented out code?

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    #op.create_unique_constraint('users_idp_userid_key', 'users', ['idp_userid'])
+    # ### end Alembic commands ###
+
+
+def downgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    #op.create_unique_constraint('users_sub_key', 'users', ['sub'])
+    #op.drop_constraint('users_idp_userid_key', 'users', ['idp_userid'])
```
Comment (seeker25) [inline legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py:30]: Any reason for the commented out code?

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py (hunk 1 above)
Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py:21]: I can update the length. And this column will be unique key later.

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py (hunk 2 above)
Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py:22]: This is just the first step. The constraint may stop the data loading step on the next.

File: legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py (hunk 1 above)
Comment (argush3) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: It's possible for it to be null though right?

Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: no, it shouldn't be null. It has to have the value from Keycloak.

Comment (argush3) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: There is user data from COLIN where we don't have that info.  So it needs to be nullable I think.

Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: What's information in the 'sub' column now of Colin user? We can copy the value from 'sub' to this column.

Comment (argush3) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: "sub" info was not available in COLIN db from what I could see.  This is the table where the user info came from.

image

Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: Do you have example in prod or test database? I need to look at the data first.

Comment (argush3) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: Here are some examples in LEAR Prod db.

I think in a lot of instances, there was a userid match and that user was just used so there was a sub.  There are 172 user entries in Prod without a sub and I think a lot of those were brought over from COLIN.

```
+----+--------+---------+--------+-----+----+----+---------------------------------+
|id  |username|firstname|lastname|email|sub |iss |creation_date                    |
+----+--------+---------+--------+-----+----+----+---------------------------------+
|2358|SC70011 |         |        |     |null|null|1995-10-27 07:00:00.000000 +00:00|
|2095|PV81928 |         |        |     |null|null|2000-01-04 08:00:00.000000 +00:00|
|2096|PY75075 |         |        |     |null|null|2011-04-08 07:00:00.000000 +00:00|
|2097|PT81928 |         |        |     |null|null|2011-01-24 08:00:00.000000 +00:00|
+----+--------+---------+--------+-----+----+----+---------------------------------+
```

Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: @bolyachevets thought?

Comment (pwei1018) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: @argush3 we should copy the username to the new column. I checked the existing database only few username got duplicated.

Comment (argush3) [reply legal-api/migrations/versions/1ba027b074a5_add_idp_userid_to_users.py]: ok, sounds good

**Review by argush3 — APPROVED** [review]

**Pull request merged**